File: apps/demo-be-java-vertx/src/test/java/com/organiclever/demojavx/integration/steps/AttachmentSteps.java
package com.organiclever.demojavx.integration.steps;

import com.organiclever.demojavx.support.AppFactory;
import com.organiclever.demojavx.support.ScenarioState;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.client.HttpResponse;
import io.vertx.ext.web.multipart.MultipartForm;
import java.util.concurrent.TimeUnit;
import org.junit.jupiter.api.Assertions;

public class AttachmentSteps {

    private static final int MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB

    private final ScenarioState state;

    public AttachmentSteps(ScenarioState state) {
        this.state = state;
    }

    @When("^alice uploads file \"([^\"]*)\" with content type \"([^\"]*)\" to POST /api/v1/expenses/\\{expenseId\\}/attachments$")
    public void aliceUploadsFileToExpense(String filename, String contentType) throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");

        byte[] content = createSampleContent(filename);
        MultipartForm form = MultipartForm.create()
                .binaryFileUpload("file", filename, Buffer.buffer(content), contentType);

        HttpResponse<Buffer> response = AppFactory.getClient()
                .post("/api/v1/expenses/" + expenseId + "/attachments")
                .bearerTokenAuthentication(token)
                .sendMultipartForm(form)
                .toCompletionStage()
                .toCompletableFuture()
                .get(5, TimeUnit.SECONDS);
        state.setLastResponse(response);
        if (response.statusCode() == 201) {
            state.setAttachmentId(response.bodyAsJsonObject().getString("id"));
        }
    }

    @When("^alice uploads file \"([^\"]*)\" with content type \"([^\"]*)\" to POST /api/v1/expenses/\\{bobExpenseId\\}/attachments$")
    public void aliceUploadsFileToBobsExpense(String filename, String contentType) throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getBobExpenseId();
        Assertions.assertNotNull(expenseId, "Bob's expense ID must be set");

        byte[] content = createSampleContent(filename);
        MultipartForm form = MultipartForm.create()
                .binaryFileUpload("file", filename, Buffer.buffer(content), contentType);

        HttpResponse<Buffer> response = AppFactory.getClient()
                .post("/api/v1/expenses/" + expenseId + "/attachments")
                .bearerTokenAuthentication(token)
                .sendMultipartForm(form)
                .toCompletionStage()
                .toCompletableFuture()
                .get(5, TimeUnit.SECONDS);
        state.setLastResponse(response);
    }

    @When("^alice uploads an oversized file to POST /api/v1/expenses/\\{expenseId\\}/attachments$")
    public void aliceUploadsOversizedFile() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");

        byte[] oversizedContent = new byte[MAX_FILE_SIZE + 1024];
        MultipartForm form = MultipartForm.create()
                .binaryFileUpload("file", "big.jpg", Buffer.buffer(oversizedContent), "image/jpeg");

        HttpResponse<Buffer> response = AppFactory.getClient()
                .post("/api/v1/expenses/" + expenseId + "/attachments")
                .bearerTokenAuthentication(token)
                .sendMultipartForm(form)
                .toCompletionStage()
                .toCompletableFuture()
                .get(10, TimeUnit.SECONDS);
        state.setLastResponse(response);
    }

    @When("^alice sends GET /api/v1/expenses/\\{expenseId\\}/attachments$")
    public void aliceSendsGetAttachments() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");

        HttpResponse<Buffer> response = AppFactory.getClient()
                .get("/api/v1/expenses/" + expenseId + "/attachments")
                .bearerTokenAuthentication(token)
                .send()
                .toCompletionStage()
                .toCompletableFuture()
                .get(5, TimeUnit.SECONDS);
        state.setLastResponse(response);
    }

    @When("^alice sends GET /api/v1/expenses/\\{bobExpenseId\\}/attachments$")
    public void aliceSendsGetBobsAttachments() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getBobExpenseId();
        Assertions.assertNotNull(expenseId, "Bob's expense ID must be set");

        HttpResponse<Buffer> response = AppFactory.getClient()
                .get("/api/v1/expenses/" + expenseId + "/attachments")
                .bearerTokenAuthentication(token)
                .send()
                .toCompletionStage()
                .toCompletableFuture()
                .get(5, TimeUnit.SECONDS);
        state.setLastResponse(response);
    }

    @When("^alice sends DELETE /api/v1/expenses/\\{expenseId\\}/attachments/\\{attachmentId\\}$")
    public void aliceSendsDeleteAttachment() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        String attachmentId = state.getAttachmentId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");
        Assertions.assertNotNull(attachmentId, "Attachment ID must be set");

        HttpResponse<Buffer> response = AppFactory.getClient()
                .delete("/api/v1/expenses/" + expenseId + "/attachments/" + attachmentId)
                .bearerTokenAuthentication(token)
                .send()
                .toCompletionStage()
                .toCompletableFuture()
                .get(5, TimeUnit.SECONDS);
        state.setLastResponse(response);
    }

    @When("^alice sends DELETE /api/v1/expenses/\\{bobExpenseId\\}/attachments/\\{attachmentId\\}$")
    public void aliceSendsDeleteBobsAttachment() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getBobExpenseId();
        String attachmentId = state.getAttachmentId();
        Assertions.assertNotNull(expenseId, "Bob's expense ID must be set");
        Assertions.assertNotNull(attachmentId, "Attachment ID must be set");

        HttpResponse<Buffer> response = AppFactory.getClient()
                .delete("/api/v1/expenses/" + expenseId + "/attachments/" + attachmentId)
                .bearerTokenAuthentication(token)
                .send()
                .toCompletionStage()
                .toCompletableFuture()
                .get(5, TimeUnit.SECONDS);
        state.setLastResponse(response);
    }

    @When("^alice sends DELETE /api/v1/expenses/\\{expenseId\\}/attachments/\\{randomAttachmentId\\}$")
    public void aliceSendsDeleteNonExistentAttachment() throws Exception {
        String token = state.getAccessToken();
        String expenseId = state.getExpenseId();
        Assertions.assertNotNull(expenseId, "Expense ID must be set");

        HttpResponse<Buffer> response = AppFactory.getClient()
                .delete("/api/v1/expenses/" + expenseId + "/attachments/99999")
                .bearerTokenAuthentication(token)
                .send()
                .toCompletionStage()
                .toCompletableFuture()
                .get(5, TimeUnit.SECONDS);
        state.setLastResponse(response);
    }

    @Given("alice has uploaded file {string} with content type {string} to the entry")
    public void aliceHasUploadedFileToEntry(String filename, String contentType) throws Exception {
        aliceUploadsFileToExpense(filename, contentType);
        HttpResponse<Buffer> resp = state.getLastResponse();
        if (resp != null && resp.statusCode() == 201) {
            state.setAttachmentId(resp.bodyAsJsonObject().getString("id"));
        }
    }

    @Given("^bob has created an entry with body \\{ \"amount\": \"([^\"]*)\", \"currency\": \"([^\"]*)\", \"category\": \"([^\"]*)\", \"description\": \"([^\"]*)\", \"date\": \"([^\"]*)\", \"type\": \"([^\"]*)\" \\}$")
    public void bobHasCreatedEntry(String amount, String currency, String category,
            String description, String date, String type) throws Exception {
        String bobToken = state.getBobAccessToken();
        Assertions.assertNotNull(bobToken, "Bob's access token must be set");
        JsonObject body = new JsonObject()
                .put("amount", amount)
                .put("currency", currency)
                .put("category", category)
                .put("description", description)
                .put("date", date)
                .put("type", type);
        HttpResponse<Buffer> resp = AppFactory.getClient()
                .post("/api/v1/expenses")
                .bearerTokenAuthentication(bobToken)
                .sendJsonObject(body)
                .toCompletionStage()
                .toCompletableFuture()
                .get(5, TimeUnit.SECONDS);
        Assertions.assertEquals(201, resp.statusCode(),
                "Expected 201 creating bob's entry but got " + resp.statusCode());
        state.setBobExpenseId(resp.bodyAsJsonObject().getString("id"));
    }

    @Then("the response body should contain {int} items in the {string} array")
    public void responseBodyContainsItemsInArray(int count, String field) {
        HttpResponse<Buffer> response = state.getLastResponse();
        Assertions.assertNotNull(response);
        JsonObject body = response.bodyAsJsonObject();
        Assertions.assertNotNull(body);
        JsonArray arr = body.getJsonArray(field);
        Assertions.assertNotNull(arr, "Expected '" + field + "' array in response");
        Assertions.assertEquals(count, arr.size(),
                "Expected " + count + " items in '" + field + "' but got " + arr.size());
    }

    @Then("the response body should contain an attachment with {string} equal to {string}")
    public void responseBodyContainsAttachmentWithField(String field, String value) {
        HttpResponse<Buffer> response = state.getLastResponse();
        Assertions.assertNotNull(response);
        JsonObject body = response.bodyAsJsonObject();
        Assertions.assertNotNull(body);
        JsonArray attachments = body.getJsonArray("attachments");
        Assertions.assertNotNull(attachments, "Expected 'attachments' array");
        boolean found = false;
        for (int i = 0; i < attachments.size(); i++) {
            JsonObject att = attachments.getJsonObject(i);
            if (value.equals(att.getString(field))) {
                found = true;
                break;
            }
        }
        Assertions.assertTrue(found,
                "Expected attachment with '" + field + "' = '" + value + "'");
    }

    private byte[] createSampleContent(String filename) {
        if (filename.endsWith(".jpg") || filename.endsWith(".jpeg")) {
            return new byte[]{(byte) 0xFF, (byte) 0xD8, (byte) 0xFF, (byte) 0xE0};
        } else if (filename.endsWith(".pdf")) {
            return "%PDF-1.4 sample".getBytes();
        } else {
            return "sample content".getBytes();
        }
    }
}

File: apps/demo-be-kotlin-ktor/src/test/kotlin/com/organiclever/demoktkt/integration/steps/AttachmentSteps.kt
package com.organiclever.demoktkt.integration.steps

import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue

class AttachmentSteps {

  private fun uploadFile(
    username: String,
    filename: String,
    contentType: String,
    expenseIdKey: String,
    fileContent: ByteArray = "test file content".toByteArray(),
  ): Pair<Int, String> {
    val token = TestWorld.accessTokens[username] ?: error("$username has no access token")
    val expenseId =
      TestWorld.expenseIds[expenseIdKey] ?: error("no expense id stored at $expenseIdKey")
    return HttpHelper.postMultipart(
      "/api/v1/expenses/$expenseId/attachments",
      filename,
      contentType,
      fileContent,
      token,
    )
  }

  @Given(
    "bob has created an entry with body \\{ {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string} \\}"
  )
  fun bobHasCreatedEntry(
    k1: String,
    v1: String,
    k2: String,
    v2: String,
    k3: String,
    v3: String,
    k4: String,
    v4: String,
    k5: String,
    v5: String,
    k6: String,
    v6: String,
  ) {
    // Login bob
    val passwords = listOf("Str0ng#Pass2", "Str0ng#Pass1")
    for (pwd in passwords) {
      val (status, body) =
        HttpHelper.post("/api/v1/auth/login", """{"username":"bob","password":"$pwd"}""")
      if (status == 200) {
        JsonHelper.getString(body, "access_token")?.let { TestWorld.accessTokens["bob"] = it }
        break
      }
    }
    val body = """{"$k1":"$v1","$k2":"$v2","$k3":"$v3","$k4":"$v4","$k5":"$v5","$k6":"$v6"}"""
    val token = TestWorld.accessTokens["bob"] ?: error("bob has no access token")
    val (status, respBody) = HttpHelper.post("/api/v1/expenses", body, token)
    assertTrue(status == 201, "Expected 201 creating bob's expense, got $status. Body: $respBody")
    JsonHelper.getString(respBody, "id")?.let { TestWorld.expenseIds["bob:last"] = it }
  }

  @Given("^alice has uploaded file \"([^\"]+)\" with content type \"([^\"]+)\" to the entry$")
  fun aliceHasUploadedFileToEntry(filename: String, contentType: String) {
    val (status, body) = uploadFile("alice", filename, contentType, "alice:last")
    assertTrue(status == 201, "Expected 201 uploading file, got $status. Body: $body")
    JsonHelper.getString(body, "id")?.let { TestWorld.attachmentIds["alice:last"] = it }
  }

  @When(
    "^alice uploads file \"([^\"]+)\" with content type \"([^\"]+)\" to POST /api/v1/expenses/\\{expenseId\\}/attachments$"
  )
  fun aliceUploadsFileToExpense(filename: String, contentType: String) {
    val (status, body) = uploadFile("alice", filename, contentType, "alice:last")
    TestWorld.lastResponseStatus = status
    TestWorld.lastResponseBody = body
    if (status == 201) {
      JsonHelper.getString(body, "id")?.let { TestWorld.attachmentIds["alice:last"] = it }
    }
  }

  @When(
    "^alice uploads file \"([^\"]+)\" with content type \"([^\"]+)\" to POST /api/v1/expenses/\\{bobExpenseId\\}/attachments$"
  )
  fun aliceUploadsToBobsExpense(filename: String, contentType: String) {
    val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val expenseId = TestWorld.expenseIds["bob:last"] ?: error("no bob expense id stored")
    val (status, body) =
      HttpHelper.postMultipart(
        "/api/v1/expenses/$expenseId/attachments",
        filename,
        contentType,
        "test file content".toByteArray(),
        token,
      )
    TestWorld.lastResponseStatus = status
    TestWorld.lastResponseBody = body
  }

  @When(
    "alice uploads an oversized file to POST \\/api\\/v1\\/expenses\\/\\{expenseId\\}\\/attachments"
  )
  fun aliceUploadsOversizedFile() {
    val oversizedContent = ByteArray(11 * 1024 * 1024) { 'X'.code.toByte() } // 11MB
    val (status, body) =
      uploadFile("alice", "large.jpg", "image/jpeg", "alice:last", oversizedContent)
    TestWorld.lastResponseStatus = status
    TestWorld.lastResponseBody = body
  }

  @When("alice sends GET \\/api\\/v1\\/expenses\\/\\{expenseId\\}\\/attachments")
  fun aliceSendsGetAttachments() {
    val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val expenseId = TestWorld.expenseIds["alice:last"] ?: error("no expense id stored")
    val (status, body) = HttpHelper.get("/api/v1/expenses/$expenseId/attachments", token)
    TestWorld.lastResponseStatus = status
    TestWorld.lastResponseBody = body
  }

  @When("^alice sends GET /api/v1/expenses/\\{bobExpenseId\\}/attachments$")
  fun aliceSendsGetBobsAttachments() {
    val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val expenseId = TestWorld.expenseIds["bob:last"] ?: error("no bob expense id stored")
    val (status, body) = HttpHelper.get("/api/v1/expenses/$expenseId/attachments", token)
    TestWorld.lastResponseStatus = status
    TestWorld.lastResponseBody = body
  }

  @When(
    "alice sends DELETE \\/api\\/v1\\/expenses\\/\\{expenseId\\}\\/attachments\\/\\{attachmentId\\}"
  )
  fun aliceSendsDeleteAttachment() {
    val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val expenseId = TestWorld.expenseIds["alice:last"] ?: error("no expense id stored")
    val attachmentId = TestWorld.attachmentIds["alice:last"] ?: error("no attachment id stored")
    val (status, body) =
      HttpHelper.delete("/api/v1/expenses/$expenseId/attachments/$attachmentId", token)
    TestWorld.lastResponseStatus = status
    TestWorld.lastResponseBody = body
  }

  @When("^alice sends DELETE /api/v1/expenses/\\{bobExpenseId\\}/attachments/\\{attachmentId\\}$")
  fun aliceSendsDeleteAttachmentOnBobsExpense() {
    val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val expenseId = TestWorld.expenseIds["bob:last"] ?: error("no bob expense id stored")
    val attachmentId =
      TestWorld.attachmentIds["alice:last"] ?: "00000000-0000-0000-0000-000000000000"
    val (status, body) =
      HttpHelper.delete("/api/v1/expenses/$expenseId/attachments/$attachmentId", token)
    TestWorld.lastResponseStatus = status
    TestWorld.lastResponseBody = body
  }

  @When(
    "^alice sends DELETE /api/v1/expenses/\\{expenseId\\}/attachments/\\{randomAttachmentId\\}$"
  )
  fun aliceSendsDeleteNonExistentAttachment() {
    val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val expenseId = TestWorld.expenseIds["alice:last"] ?: error("no expense id stored")
    val randomId = "00000000-0000-0000-0000-000000000000"
    val (status, body) =
      HttpHelper.delete("/api/v1/expenses/$expenseId/attachments/$randomId", token)
    TestWorld.lastResponseStatus = status
    TestWorld.lastResponseBody = body
  }

  @Then("the response body should contain {int} items in the {string} array")
  fun theResponseBodyShouldContainItemsInArray(count: Int, arrayKey: String) {
    val actual = JsonHelper.getArraySize(TestWorld.lastResponseBody, arrayKey)
    assertEquals(
      count,
      actual,
      "Expected $count items in '$arrayKey' array in: ${TestWorld.lastResponseBody}",
    )
  }

  @Then("the response body should contain an attachment with {string} equal to {string}")
  fun theResponseBodyShouldContainAttachmentWithField(field: String, expected: String) {
    val body = TestWorld.lastResponseBody
    assertTrue(body.contains(expected), "Expected attachment with '$field'='$expected' in: $body")
  }
}
